### PR TITLE
Retains tree state of function tree during Sort/Unsort and Save/Reload actions

### LINKED ISSUE #8819: Function List: Current entry is lost while unsorting/refreshing
Reading about Function List issues in #8787 and #8796, I cannot refrain from putting this one into the discussion too.

### Description of the Issue
The current function list entry is lost while unsorting/refreshing

### Steps to Reproduce the Issue

1.  Open Notepad_plus.cpp in Notepad++
2.  Open the Function List panel.
3. Click "init" in the Function List panel.
4. Click the sort button in the Function List Panel
5. Click the sort button in the Function List Panel again.

### Expected Behavior
After step 4, the selected entry in the Function List Panel is "init".
After step 5, the selected entry in the Function List Panel is still "init".

### Actual Behavior
After step 4: as expected
After step 5: Folding state and position is lost.

Similar effects occur when a search string is entered in the edit window of the Function List, or when a sorted view of the Function List is being refreshed.

### Debug Information
Use a freshly unpacked `npp.7.8.9.bin.7z` to reproduce the issue.

Notepad++ v7.8.9   (32-bit)
Build time : Jul 15 2020 - 20:26:50
Path : C:\Tools\Npp_789_t\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Pro (64-bit) 
OS Version : 1607
OS Build : 14393.576
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll NppExport.dll 

Ups, I'm surprised, that there seem to be pre-installed Plugins here, but I don't think, that they have anything to do with the described effects.

### LINKED ISSUE #10280: Function List: retain tree state after Save/Reload
**STR 1:**
Open `Notepad_plus.cpp`.
Open the Function List panel.
You get the following state.

תמונה

Double-click isUrl, edit that function body, save the file and return focus to the FL by clicking on its caption.

**Result:** 
The root `Notepad_plus.cpp` node is selected.

**Expected:**
`isUrl` should be selected.

***
**STR 2:**
Open `Notepad_plus.cpp`.
Open the Function List panel.
Expand the `Notepad_plus` node.
You get the following state.

תמונה

Press `Sort by Name`.
Double-click `replaceInFiles`, edit that function body, save and return focus to the FL by clicking on its caption.

**Result:** 
The tree is collapsed and the `Notepad_plus.cpp` node is selected.

**Expected:**
The tree should retain its expanded state and `replaceInFiles` should be selected.
***
Notepad++ v8.1.2   (32-bit)
Build time : Jul 28 2021 - 01:28:03
Path : C:\---\notepad++.exe
Command Line : 
Admin mode : ON
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 2009
OS Build : 19042.985
Current ANSI codepage : 1255
Plugins : ComparePlus.dll HTMLTag_unicode.dll MenuIcons.dll PythonScript.dll _CustomizeToolbar.dll

### LINKED ISSUE #10294: Function List: retain tree state after Sort/Unsort
### Description of the Issue

The functions list's expanded sub functions collapse when unsorting a sorted function list. The gif below demonstrates the problem

funcListTreeSortProb

### Steps to Reproduce the Issue

1.  Open a file along with the function list pane, make sure the function list has toggleable dropdowns
2.  expand the dropdowns.
3.  Click the sort button twice (Sort and then unsort)

### Expected Behavior

retain the expanded state of the sub functions in the function list after the 2nd sort button click

### Actual Behavior

function list's sub expanded functions collapse after the 2nd sort button click

### Debug Information

Notepad++ v8.1.2   (64-bit)
Build time : Jul 16 2021 - 01:59:50
Path : C:\Users\ASHFAAQ\Downloads\npp.8.1.2.portable.x64\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit) 
OS Version : 2009
OS Build : 19042.1110
Current ANSI codepage : 1252
Plugins : mimeTools.dll NppConverter.dll NppExport.dll

### PR DESCRIPTION
**Pull request by Ashfaaq18**
### fix #10280, fix #10294 and fix #8819

below 2 gifs for fix #10280 

STR1 fix:
funcListTreeSortProb-Str1

STR2 fix:
funcListTreeSortProb-Str2

below gif for fix #10294

funcListTreeSortProb-Str3

### CODE REVIEW

**Comment by Yaron10** [comment]
Hello @Ashfaaq18,

Brilliant! 👍 
Thank you for your work.

Some comments.

***
Have the tree in the following state.

תמונה

Press "Sort".

**Result:**

תמונה

(The horizontal scroll bar moved to the right.
***

Expand the tree.
Select a node and scroll so that it's not visible.

On "Reload", the scroll bar position is retained and the selected node is still not visible.
On "Sort/Unsort", the selected node becomes visible and the scroll bar position is overridden.

IMO - in both cases the selected node should become visible.
(This is the behavior in File Explorer: select an item, scroll so that it's not visible and Refresh - the item becomes visible).
If trying to center the selected node would not require a lot of work/code - it might be more convenient.

***
This is not directly related to this PR.

Have the caret in the editor in function A.
**Single-click** function B in the FL.
Minimize NPP and restore.

**Result:**
The editor gets the focus and function A is selected in the FL.

**Comment by Ashfaaq18** [comment]
Hi @Yaron10 , thanks for the feedback. Ill get back to your comments once @donho takes a look at this PR

**Comment by ozone10** [comment]
@Ashfaaq18 
Your branch is several commit behind, could you rebase it on newer master.
There were changes in some files you are modifying, this could result in conflict.

**Comment by Yaron10** [comment]
@Ashfaaq18,

Great.
Thanks again.

File: PowerEditor/src/WinControls/TreeView/TreeView.cpp [hunk 1]
```diff
@@ -670,31 +670,14 @@ bool TreeView::restoreFoldingStateFrom(const TreeStateNode & treeState2Compare,
...
-
 	if (treeState2Compare._isExpanded) //= (tvItem.state & TVIS_EXPANDED) != 0;
 		expand(treeviewNode);
 	else
 		fold(treeviewNode);
 
 	if (treeState2Compare._isSelected) //= (tvItem.state & TVIS_SELECTED) != 0;
 		selectItem(treeviewNode);
-
+
```
Comment (donho) [inline PowerEditor/src/WinControls/TreeView/TreeView.cpp:680]: Please remove this diff.

**Review by donho — CHANGES_REQUESTED** [review]
Remove unnecessary diff.

**Review by donho — APPROVED** [review]

**Comment by donho** [comment]
@Ashfaaq18 
This PR is approved and it'll be merged into master in about 1 week.
The master branch is frozen currently, the RC is being tested by the community:
https://community.notepad-plus-plus.org/topic/21599/notepad-v8-1-3-release-candidate

**Pull request closed without merging**